### ISSUE TITLE
Wrong type information for `firebase.storage.UploadTask.on`

### ISSUE THREAD

**Issue by henriiik** [firebase/firebase-js-sdk#620]
### [REQUIRED] Describe your environment

  * Operating System version: macOS 10.13.3
  * Firebase SDK version: 4.12.1
  * Firebase Product: storage

### [REQUIRED] Describe the problem

The when registering an observer on an upload task, the `next` function is specified as having one argument of type `Object` but it is `UploadTaskSnapshot`. This leads to errors when accessing it's properties.

#### Steps to reproduce:

```ts
import * as firebase from "firebase";

const task = firebase
	.storage()
	.ref("/hello.txt")
	.putString("hello");

task.on(firebase.storage.TaskEvent.STATE_CHANGED, snapshot => {
	var progress = snapshot.bytesTransferred / snapshot.totalBytes * 100;
	console.log("Upload is " + progress + "% done");
});
```
This minimal example (adapted from the docs https://firebase.google.com/docs/storage/web/upload-files#monitor_upload_progress) results in the following error:

```
(9,26): Property 'bytesTransferred' does not exist on type 'Object'.
```

The typings of the project say that the snapshot variable is of type `Object` but it is of type `firebase.storage.UploadTaskSnapshot`.

If you try to update the type of the callback:

```ts
import * as firebase from "firebase";

const task = firebase
	.storage()
	.ref("/hello.txt")
	.putString("hello");

task.on(
	firebase.storage.TaskEvent.STATE_CHANGED,
	(snapshot: firebase.storage.UploadTaskSnapshot) => {
		var progress = snapshot.bytesTransferred / snapshot.totalBytes * 100;
		console.log("Upload is " + progress + "% done");
	}
);
```

you are given another error (in --strict mode):

```
(10,2): Argument of type '(snapshot: UploadTaskSnapshot) => void' is not assignable to parameter of type 'Observer<any, any> | ((a: Object) => any) | null | undefined'.
  Type '(snapshot: UploadTaskSnapshot) => void' is not assignable to type '(a: Object) => any'.
    Types of parameters 'snapshot' and 'a' are incompatible.
      Type 'Object' is not assignable to type 'UploadTaskSnapshot'.
        The 'Object' type is assignable to very few other types. Did you mean to use the 'any' type instead?
          Property 'bytesTransferred' is missing in type 'Object'.
```

#### Relevant Code:

This is the relevant code for the real type type in the source: https://github.com/firebase/firebase-js-sdk/blob/4f280965a2f285e4eecb88ade276747031bf9769/packages/storage/src/task.ts#L505

This is the relevant code for the exported type: https://github.com/firebase/firebase-js-sdk/blob/4f280965a2f285e4eecb88ade276747031bf9769/packages/storage-types/index.d.ts#L80

#### Possible Fix:

I have created a small fix that would solve this problem: https://github.com/henriiik/firebase-js-sdk/commit/f5be6543f240d1eb1e78287a0a98e58422a7fb62

**Comment by henriiik** [comment]
I mistakenly opened this issue before I was done writing it, so I closed it. It's now complete and so I open it again. :)

**Comment by schmidt-sebastian** [comment]
Thanks for sending this over. We should be able to address this shortly.

**Comment by sphippen** [comment]
Fixed in #642.
